kinto/core/views/heartbeat.py

```diff
@@ -11,10 +14,28 @@ def get_heartbeat(request):
     """Return information about server health."""
     status = {}
 
-    heartbeats = request.registry.heartbeats
-    for name, callable in heartbeats.items():
-        status[name] = callable(request)
+    def heartbeat_check(name, func):
+        status[name] = False
+        status[name] = func(request)
 
+    # Start executing heartbeats concurrently.
+    heartbeats = request.registry.heartbeats
+    pool = ThreadPoolExecutor(max_workers=len(heartbeats.keys()))
+    futures = []
+    for name, func in heartbeats.items():
+        future = pool.submit(heartbeat_check, name, func)
+        future.__heartbeat_name = name  # For logging purposes.
+        futures.append(future)
+
+    # Wait the results, with timeout.
+    seconds = float(request.registry.settings['heartbeat_timeout_seconds'])
+    done, not_done = wait(futures, timeout=seconds)
+    if len(not_done) > 0:
+        name = not_done.pop().__heartbeat_name
+        error_msg = "'%s' heartbeat has exceeded timeout of %s seconds."
+        logger.error(error_msg % (name, seconds))
+
+    # If any has failed, return a 503 error response.
```

Comment:
also: make sure you call `logger.exception` in the main thread sequentially on the collected tracebacks, otherwise you might have mangled exceptions since two functions can fail in parallel at the same instant